### PR TITLE
Embed attribute mapping into entity descriptor

### PR DESCRIPTION
**Pull request by flyinghermit**
Embeds attribute mapping details provided in `saml_idp_service_provider` spec into Service provider entity descriptor.

attribute mapping input:
```
sp.yml
kind: saml_idp_service_provider
metadata:
  name: sp
spec:
  attribute_mapping:
   - name: groups
      value: user.spec.traits.groups
```

becomes xml entity SPSSODescriptor:
```
<EntityDescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" validUntil="2025-12-09T09:13:31.006Z" entityID="sp">
         <SPSSODescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol" AuthnRequestsSigned="false" WantAssertionsSigned="true">
             <NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</NameIDFormat>
             <AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://example.com/acs" index="0" isDefault="true"></AssertionConsumerService>
         </SPSSODescriptor>
         <SPSSODescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" protocolSupportEnumeration="">
             <AttributeConsumingService index="0">
                 <ServiceName xml:lang="">teleport_saml_idp_service</ServiceName>
                 <RequestedAttribute FriendlyName="groups" Name="groups" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
                     <AttributeValue xmlns:_XMLSchema-instance="http://www.w3.org/2001/XMLSchema-instance" _XMLSchema-instance:type="">user.spec.traits.groups</AttributeValue>
                 </RequestedAttribute>
             </AttributeConsumingService>
         </SPSSODescriptor>
</EntityDescriptor>
```

While embedding, we add a `ServiceName` field with value `teleport_saml_idp_service`. This will let us properly identify embedded attributes from other entity descriptor elements. 

RFD: https://github.com/gravitational/teleport.e/pull/2801
see https://github.com/gravitational/teleport/issues/34725, https://github.com/gravitational/teleport.e/issues/2692

### CODE REVIEW

File: lib/services/local/saml_idp_service_provider.go [hunk 1]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
 	sp.SetEntityDescriptor(string(entityDescriptor))
 	return nil
 }
+
+// embedAttributeMapping embeds attribute mapping details into entity descriptor.
+func (s *SAMLIdPServiceProviderService) embedAttributeMapping(sp types.SAMLIdPServiceProvider) error {
+	if len(sp.GetAttributeMapping()) == 0 {
+		s.log.Infof("no custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:314]:
```suggestion
		s.log.Debugf("No custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
```

File: lib/services/local/saml_idp_service_provider.go [hunk 2]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+	} else {
+		ed.SPSSODescriptors[embeddedSPSSODescriptorIndex] = newEmbeddedSPSSODescriptor
+	}
+
+	edWithAttributes, err := xml.MarshalIndent(ed, " ", "    ")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.log.Infof("attribute mapping embedded in entity descriptor")
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:344]: I would just remove this log entry, doesn't seem useful.
```suggestion
```

File: lib/services/local/saml_idp_service_provider.go [hunk 3]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+	ed, err := samlsp.ParseMetadata([]byte(sp.GetEntityDescriptor()))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// return if provided attributes are already embedded.
+	embeddedSPSSODescriptorIndex, embeddedAttributes := samlSPSSODescriptorToProtoSAMLAttributeMapping(ed.SPSSODescriptors)
+	if reflect.DeepEqual(sp.GetAttributeMapping(), embeddedAttributes) {
+		return nil
+	}
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:327]: I've a couple of questions about this block:
- In what case will the attributes be already embedded?
- Is "deep equal" check appropriate here? What if the attribute mapping already has a superset of attributes?

File: lib/services/local/saml_idp_service_provider.go [hunk 4]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+					embeddedSPSSODescriptorIndex = descriptorIndex
+					for _, reqAttr := range acs.RequestedAttributes {
+						var embeddedAttribute types.SAMLAttributeMapping
+						for _, reqAttrVal := range reqAttr.Values {
+							embeddedAttribute.Name = reqAttr.Name
+							embeddedAttribute.NameFormat = reqAttr.NameFormat
+							embeddedAttribute.Value = reqAttrVal.Value
+						}
+						attrs = append(attrs, &embeddedAttribute)
+					}
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:398]: Should we return here? Or can there be multiple descriptors for "teleport" service name?

File: lib/services/local/saml_idp_service_provider.go [hunk 5]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+				if serviceName.Value == teleportSAMLIdP {
+					embeddedSPSSODescriptorIndex = descriptorIndex
+					for _, reqAttr := range acs.RequestedAttributes {
+						var embeddedAttribute types.SAMLAttributeMapping
+						for _, reqAttrVal := range reqAttr.Values {
+							embeddedAttribute.Name = reqAttr.Name
+							embeddedAttribute.NameFormat = reqAttr.NameFormat
+							embeddedAttribute.Value = reqAttrVal.Value
+						}
+						attrs = append(attrs, &embeddedAttribute)
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:397]: This can be written slightly more efficiently, without an extra variable:

```go
for _, reqAttrVal := range reqAttr.Values {
	attrs = append(attrs, &types.SAMLAttributeMapping{
		Name: reqAttr.Name,
		NameFormat: reqAttr.NameFormat,
		Value: retAttr.Value,
	}
}
```

File: lib/services/local/saml_idp_service_provider.go [hunk 6]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+// and returns SPSSODescriptor element index and converted attribute mapping values.
+// The correct SPSSODescriptor is determined by searching for AttributeConsumingService element with ServiceNames named
+// TELEPORT_SAML_IDP.
+func samlSPSSODescriptorToProtoSAMLAttributeMapping(spSSODescriptors []saml.SPSSODescriptor) (int, []*types.SAMLAttributeMapping) {
+	embeddedSPSSODescriptorIndex := 0
+	attrs := make([]*types.SAMLAttributeMapping, 0)
+	for descriptorIndex, descriptor := range spSSODescriptors {
+		for _, acs := range descriptor.AttributeConsumingServices {
+			for _, serviceName := range acs.ServiceNames {
+				if serviceName.Value == teleportSAMLIdP {
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:388]: Instead of building the staircase of loops in this function, let's make a helper method that returns Teleport's descriptor and index given a list of descriptors.

File: lib/services/local/saml_idp_service_provider.go [hunk 7]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+	var reqs []saml.RequestedAttribute
+	for _, v := range attributeMapping {
+		var reqAttrs saml.RequestedAttribute
+		reqAttrs.FriendlyName = v.Name
+		reqAttrs.Name = v.Name
+		reqAttrs.NameFormat = v.NameFormat
+		reqAttrs.Values = append(reqAttrs.Values, saml.AttributeValue{
+			Value: v.Value,
+		})
+		reqs = append(reqs, reqAttrs)
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:361]: You don't need an extra variable, this will be more Go-idiomatic:

```go
reqs = append(reqs, saml.RequestedAttribute{
  FriendlyName: v.Name,
  Name: v.Name,
  ....
})
```

File: lib/services/local/saml_idp_service_provider.go [hunk 8]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+	}
+
+	// return if provided attributes are already embedded.
+	embeddedSPSSODescriptorIndex, embeddedAttributes := samlSPSSODescriptorToProtoSAMLAttributeMapping(ed.SPSSODescriptors)
+	if reflect.DeepEqual(sp.GetAttributeMapping(), embeddedAttributes) {
+		return nil
+	}
+
+	newEmbeddedSPSSODescriptor := spSSODescriptorWithRequestedAttributes(sp.GetAttributeMapping())
+	// If there is an existing embedded SPSSODescriptor, we'll replace it with newEmbeddedSPSSODescriptor
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:330]: Nit: I would rephrase the comment a bit to clarify that if there's an existing descriptor for "teleport" service, we update and replace it, otherwise we create a new one for "teleport" service and append it. It took me a bit to figure this out.

File: lib/services/local/saml_idp_service_provider.go [hunk 9]
```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+	embeddedSPSSODescriptorIndex := 0
+	attrs := make([]*types.SAMLAttributeMapping, 0)
+	for descriptorIndex, descriptor := range spSSODescriptors {
+		for _, acs := range descriptor.AttributeConsumingServices {
+			for _, serviceName := range acs.ServiceNames {
+				if serviceName.Value == teleportSAMLIdP {
+					embeddedSPSSODescriptorIndex = descriptorIndex
+					for _, reqAttr := range acs.RequestedAttributes {
+						var embeddedAttribute types.SAMLAttributeMapping
+						for _, reqAttrVal := range reqAttr.Values {
```
Comment (mdwn) [inline lib/services/local/saml_idp_service_provider.go:392]: Won't this loop continually overwrite the embedded attribute if there are multiple values? Is the intent here to only take the last value?

File: lib/services/local/saml_idp_service_provider.go (hunk 4 above)
Comment (flyinghermit) [reply lib/services/local/saml_idp_service_provider.go:398]: Actually yes we should return there. There will always be one descriptor with "teleport" service name.

File: lib/services/local/saml_idp_service_provider.go (hunk 3 above)
Comment (flyinghermit) [reply lib/services/local/saml_idp_service_provider.go:327]: > In what case will the attributes be already embedded?

example:
1. user creates sp with entity descriptor and attribute mapping. initial sets of attributes are embedded. 
2. Then, user wants to update attribute, does `tctl get saml_idp_service_provider/sp > sp.yml`, adds new attributes to `sp.yml` and then updates. At this point, since there will be a set of attributes already embedded in entity descriptor (from step 1). If we do not check for that, we risk duplicate them. (that's why go through loop to find index)

> Is "deep equal" check appropriate here? What if the attribute mapping already has a superset of attributes?

So while I was working, I noticed `CreateSAMLIdPServiceProvider` and `UpdateSAMLIdPServiceProvider` were invoked more times than I initially would have thought (during backend insert, during cache upsert) so I thought if it would be good to break out from embedding if entity descriptor already has all the attributes mapped, hence used deep equal here to compare whats in attribute mapping vs. entity descriptor.

File: lib/services/local/saml_idp_service_provider.go (hunk 9 above)
Comment (flyinghermit) [reply lib/services/local/saml_idp_service_provider.go:392]: Yes, we do not support multiple values, only one per attribute.

File: lib/services/local/saml_idp_service_provider.go (hunk 8 above)
Comment (flyinghermit) [reply lib/services/local/saml_idp_service_provider.go]: updated in https://github.com/gravitational/teleport/pull/35705/commits/806e5625965fcc580d7c851e2df7f25fd042d7d5. let me know how it reads now.

File: lib/services/local/saml_idp_service_provider.go (hunk 6 above)
Comment (flyinghermit) [reply lib/services/local/saml_idp_service_provider.go:386]: While I don't like it myself, I am not sure if we could get away with that nested loop entirely (at least from what I can think of my head🤔) cause the type `[]saml.SPSSODescriptor` comes from external package and all of it is nested slices

**Comment by flyinghermit** [comment]
PTAL again @r0mant @mdwn

File: lib/services/local/saml_idp_service_provider.go [hunk 10]
```diff
@@ -46,6 +46,7 @@ const (
 	samlIDPServiceProviderModifyLock    = "saml_idp_service_provider_modify_lock"
 	samlIDPServiceProviderModifyLockTTL = time.Second * 5
 	samlIDPServiceProviderMaxPageSize   = 200
+	teleportSAMLIdP                     = "TELEPORT_SAML_IDP"
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:49]: Nit: I think the purpose of this constant would be slightly clearer with this name.
```suggestion
	samlIDPServiceName                  = "TELEPORT_SAML_IDP"
```

File: lib/services/local/saml_idp_service_provider.go [hunk 11]
```diff
@@ -295,3 +306,82 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+// list of SPSSODescriptors. The correct SPSSODescriptor is determined by searching for
+// AttributeConsumingService element with ServiceNames named TELEPORT_SAML_IDP.
+func GetTeleportSPSSODescriptor(spSSODescriptors []saml.SPSSODescriptor) (embeddedSPSSODescriptorIndex int, teleportSPSSODescriptor saml.SPSSODescriptor) {
+	for descriptorIndex, descriptor := range spSSODescriptors {
+		for _, acs := range descriptor.AttributeConsumingServices {
+			for _, serviceName := range acs.ServiceNames {
+				if serviceName.Value == teleportSAMLIdP {
+					embeddedSPSSODescriptorIndex = descriptorIndex
+					teleportSPSSODescriptor = spSSODescriptors[descriptorIndex]
+					return
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:352]:
```suggestion
					return descriptorIndex, spSSODescriptors[descriptorIndex]
```

File: lib/services/local/saml_idp_service_provider.go [hunk 12]
```diff
@@ -295,3 +306,82 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+func (s *SAMLIdPServiceProviderService) embedAttributeMapping(sp types.SAMLIdPServiceProvider) error {
+	ed, err := samlsp.ParseMetadata([]byte(sp.GetEntityDescriptor()))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	existingTeleportSPSSODescriptorIndex, _ := GetTeleportSPSSODescriptor(ed.SPSSODescriptors)
+	switch atrrMapLen := len(sp.GetAttributeMapping()); {
+	case atrrMapLen == 0 && existingTeleportSPSSODescriptorIndex == 0:
+		s.log.Debugf("no custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:320]:
```suggestion
		s.log.Debugf("No custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
```

File: lib/services/local/saml_idp_service_provider.go (hunk 3 above)
Comment (flyinghermit) [reply lib/services/local/saml_idp_service_provider.go]: fyi, I have updated this and now excludes `reflect.DeepEqual` and only return early if attribute mapping input and teleport managed entity descriptor is empty. It also now handles attribute mapping deletion from the entity descriptor if it is removed from the `saml_idp_service_provider` spec, which was not handled previously.

File: lib/services/local/saml_idp_service_provider.go [hunk 13]
```diff
@@ -295,3 +306,82 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+	case atrrMapLen == 0 && existingTeleportSPSSODescriptorIndex > 0:
+		// delete Teleport SPSSODescriptor
+		ed.SPSSODescriptors = append(ed.SPSSODescriptors[:existingTeleportSPSSODescriptorIndex], ed.SPSSODescriptors[existingTeleportSPSSODescriptorIndex+1:]...)
+	case atrrMapLen > 0 && existingTeleportSPSSODescriptorIndex == 0:
+		ed.SPSSODescriptors = append(ed.SPSSODescriptors, genTeleportSPSSODescriptor(sp.GetAttributeMapping()))
+	case atrrMapLen > 0 && existingTeleportSPSSODescriptorIndex > 0:
+		// if there is existing SPSSODescriptor with "TELEPORT_SAML_IDP" service name, replace it at
+		// existingTeleportSPSSODescriptorIndex to avoid duplication or possible fragmented SPSSODescriptor.
+		ed.SPSSODescriptors[existingTeleportSPSSODescriptorIndex] = genTeleportSPSSODescriptor(sp.GetAttributeMapping())
+	}
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:331]: I think switch statement hurts readability here. How about a simple "if" with comments explaining what each case is?

```go
attrMapLen := len(sp.GetAttributeMapping())

if attrMapLen == 0 { // Service provider doesn't have custom attribute mapping configured.
  if existingTeleportSPSSODescriptorIndex == 0 {     // No existing Teleport descriptor, nothing to do.
    return nil
  } else {     // Descriptor has Teleport attributes, remove it.
    ...
  }
} else { // Service provider has custom attributes mapping
  if existingTeleportSPSSODescriptorIndex == 0 {     // Teleport descriptor doesn't exist, add it.
    ...
  } else {     // Teleport descriptor already exists, update it.
    ...
  }
}
```

File: lib/services/local/saml_idp_service_provider.go [hunk 14]
```diff
@@ -295,3 +306,82 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
...
+
+// embedAttributeMapping embeds attribute mapping input into entity descriptor.
+func (s *SAMLIdPServiceProviderService) embedAttributeMapping(sp types.SAMLIdPServiceProvider) error {
+	ed, err := samlsp.ParseMetadata([]byte(sp.GetEntityDescriptor()))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	existingTeleportSPSSODescriptorIndex, _ := GetTeleportSPSSODescriptor(ed.SPSSODescriptors)
+	switch atrrMapLen := len(sp.GetAttributeMapping()); {
```
Comment (r0mant) [inline lib/services/local/saml_idp_service_provider.go:318]: Typo:
```suggestion
	switch attrMapLen := len(sp.GetAttributeMapping()); {
```

**Review by r0mant — APPROVED** [review]
lgtm with some stylistic suggestions

File: lib/services/local/saml_idp_service_provider.go (hunk 13 above)
Comment (flyinghermit) [reply lib/services/local/saml_idp_service_provider.go:331]: Yeah right. But not really fan of deep nested if else either. I think the culprit is `existingTeleportSPSSODescriptorIndex` which actually can be `teleportSPSSODescriptorIndex` since this is the only value in `embedAttributeMapping()`.
I came up with this one, looks a bit better than previous:
```
switch attrMapLen := len(sp.GetAttributeMapping()); {
	case attrMapLen == 0:
		if teleportSPSSODescriptorIndex == 0 {
			s.log.Debugf("No custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
		} else {
			// delete Teleport SPSSODescriptor
			ed.SPSSODescriptors = append(ed.SPSSODescriptors[:teleportSPSSODescriptorIndex], ed.SPSSODescriptors[teleportSPSSODescriptorIndex+1:]...)
		}
	case attrMapLen > 0:
		if teleportSPSSODescriptorIndex == 0 {
			ed.SPSSODescriptors = append(ed.SPSSODescriptors, genTeleportSPSSODescriptor(sp.GetAttributeMapping()))
		} else {
			// if there is existing SPSSODescriptor with "teleport_saml_idp_service" service name, replace it at
			// existingTeleportSPSSODescriptorIndex to avoid duplication or possible fragmented SPSSODescriptor.
			ed.SPSSODescriptors[teleportSPSSODescriptorIndex] = genTeleportSPSSODescriptor(sp.GetAttributeMapping())
		}
	}
```

**Comment by flyinghermit** [comment]
friendly ping @mdwn

**Review by mdwn — APPROVED** [review]
